From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: [PATCH] introduce get_clock_monotonic
References: bnc#532063,LTC#55526

From: Heiko Carstens <heiko.carstens@de.ibm.com>

Introduce get_clock_monotonic() function which can be used to get a
(fast) timestamp. Resolution is the same as for get_clock(). The
only difference is that the timestamps are monotonic and don't jump
backward or forward.

Signed-off-by: Heiko Carstens <heiko.carstens@de.ibm.com>
Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>

Acked-by: John Jolly <jjolly@suse.de>
---

 arch/s390/include/asm/timex.h |   16 ++++++++++++++++
 arch/s390/kernel/time.c       |    3 ++-
 2 files changed, 18 insertions(+), 1 deletion(-)

Index: linux-sles11/arch/s390/include/asm/timex.h
===================================================================
--- linux-sles11.orig/arch/s390/include/asm/timex.h
+++ linux-sles11/arch/s390/include/asm/timex.h
@@ -85,4 +85,20 @@ int get_sync_clock(unsigned long long *c
 void init_cpu_timer(void);
 unsigned long long monotonic_clock(void);
 
+extern u64 jiffies_timer_cc;
+
+/**
+ * get_clock_monotonic - returns current time in clock rate units
+ *
+ * The caller must ensure that preemption is disabled.
+ * The clock and sched_clock_base get changed via stop_machine.
+ * Therefore preemption must be disabled when calling this
+ * function, otherwise the returned value is not guaranteed to
+ * be monotonic.
+ */
+static inline unsigned long long get_clock_monotonic(void)
+{
+	return get_clock_xt() - jiffies_timer_cc;
+}
+
 #endif
Index: linux-sles11/arch/s390/kernel/time.c
===================================================================
--- linux-sles11.orig/arch/s390/kernel/time.c
+++ linux-sles11/arch/s390/kernel/time.c
@@ -63,7 +63,8 @@
 
 static ext_int_info_t ext_int_info_cc;
 static ext_int_info_t ext_int_etr_cc;
-static u64 jiffies_timer_cc;
+u64 jiffies_timer_cc;
+EXPORT_SYMBOL(jiffies_timer_cc);
 
 static DEFINE_PER_CPU(struct clock_event_device, comparators);
 
